### DISCUSSION TITLE
Rename the `Decimal` type to `Float`?

### DISCUSSION THREAD

**Discussion by radeusgd** [enso-org/enso#6858] [Work In Progress - Libraries] [answered]
The [discussion](https://github.com/enso-org/enso/pull/6817#discussion_r1205810671) on #6817 showed that our number type names and `Value_Type` are misaligned:

> Unfortunately with `Value_Type` with have a bit of inconsistency between it and the Enso types.
> 
> <table>
> <tr>
>  <th> Value Type
>  <th> Enso Type
> <tr>
>  <td> Value_Type.Integer
>  <td> Integer
> <tr>
>  <td> Value_Type.Float
>  <td> Decimal
> <tr>
>  <td> Value_Type.Decimal
>  <td> (currently not supported) BigDecimal
> </table>
> 
> The names for value type were chosen to be close to the SQL type names. Maybe we should rename Enso `Decimal` into `Float`?? It is indeed a bit confusing to have this inconsistency. Also indeed, `Decimal` often suggests precise decimals, where we have IEEE-754 floating points.
> 

Indeed `Decimal` suggests precise decimals which is currently not the case, as Enso uses IEEE-754 floats. We shouldn't rename `Value_Type` as it does follow the commonly known conventions of _decimal_ for precise numbers and _float_ for imprecise ones and it is consistent with SQL naming (which was its purpose).

But now, `Value_Type` and the Enso numeric types are misaligned, and as one can see in the table in a pretty confusing way - both have a `Decimal` type but it means different things.

We should do something to avoid the confusion.

Shall we rename the Enso `Decimal` type to `Float`? That would make it consistent with `Value_Type` and more precisely reflect that it is IEEE-754 floating points and not precise decimals.

cc: @jdunkerley @wdanilo @JaroslavTulach

**Comment by GregoryTravis** [comment]
Or Double, since that's what it is?

**Comment by JaroslavTulach** [comment]
Remove `Integer` and `Decimal`, keep just `Number`? That would at least simplify our inheritance hierarchy!

There is no inheritance in Enso - every object is of its type and of type `Any`. Except numbers are special. There is no reason for engine to expose its internal implementations of numbers to user. We could call all our internal variants `Number`.

**Reply by radeusgd** [reply]
We need a way to distinguish whole numbers from fractional numbers.

A multitude of functions can only accept an `Integer`, that's a thing that we should be able to express in the type IMO.

**Reply by radeusgd** [reply]
Also, if we do this, there is no way to tell reliably if I'm currently working with **precise** integers or **approximating** floats - and sometimes I want to guarantee the precision. Currently, I can guarantee this by working with only `Integer` type. If we merge the types, I don't know if the precise or approximating mechanism is used under the hood.

**Reply by radeusgd** [reply]
If anything, I'd rather just remove the `Number` type and keep separate types `Integer` and 'whatever name comes from this discussion'.

After all, `Number = Integer | Decimal` - it's currently a special type, but since a long time there were plans to allow 'naming' sum types and then Number would just be an alias for this sum type.

**Reply by GregoryTravis** [reply]
Agreed, we need this type hierarchy; even languages with no type hierarchy (C) have numeric type hierarchies of some kind.

**Reply by JaroslavTulach** [reply]
C needs this _some kind of numeric type hierarchy_ to **be fast**. Dynamic languages (like JavaScript) don't need that and are OK with single `Number` type. Internally we of course have to speculate on particular type of value, but there is no need to get such hint from programmers.

**Reply by radeusgd** [reply]
> Internally we of course have to speculate on particular type of value, but there is no need to get such hint from programmers.

But that is just the 'efficiency' part. What about correctness? Many functions make sense only on integers (like indexing an array). Don't we want to be able to express this in the type?

**Comment by wdanilo** [comment]
There are multiple things to consider here:
1. The simplicity for the user. From this perspective, the Jaroslav's idea to keep just `Number` is most beautiful, but it will not work, as it is unfortunately important for the user to know what number precision they use when working with databases or data frames. Please confirm this with James and Ned. If we could have just a single `Number` type, I would be the happiest man alive, but probably its not the case.
2. The current name `Decimal` was chosen because of point 1, to keep it simple for the user, because the name "float" does not tell anyone who is not computer scientist anything sensible. However, as we need to be synchronized with databases types, this probably was wrong choice, so we should use the same names as databases use.
3. Regarding "double", I hope it was a joke :D

**Reply by farmaazon** [reply]
> 3. Regarding "double", I hope it was a joke :D

Of course, it was. `f64` is the only answer.

**Reply by jdunkerley** [reply]
We need to keep the difference between `Integer` and `Decimal` (or `Float` if renamed).

`Float` is a common name for DB users (as that is the SQL standard). 
`Decimal` implies an accurate base 10 representation which we don't have (and again is a SQL term).

**Reply by wdanilo** [reply, marked as answer]
In such a case, I believe that `Float` is the best name here (unfortunately).

**Comment by jdunkerley** [comment]
@radeusgd - lets add a ticket to do this please.
